### ISSUE TITLE
Saxy hangs the dialyzer

### ISSUE THREAD

**Issue by the-mikedavis** [qcam/saxy#39]
Adding saxy as a dependency will hang the dialyzer during the deps adding stage. Running the dialyzer on saxy itself hangs during the project-checking stage.

By removing `lib/saxy/parser/{element,prolog}.ex`, the dialyzer finishes very quickly.

I would guess that this is related to the large guard clauses defined in `lib/saxy/guards.ex`.

[Here's](https://github.com/elixir-lang/elixir/pull/7993) a similar issue from the Elixir kernel. See also [this erlang bug report](https://bugs.erlang.org/browse/ERL-680). It could be that there's still an issue with large `or` guards.

I have tried Elixir and Erlang versions:

- 1.6.5 OTP 20.3
- 1.7.4 OTP 21.0
- 1.7.4 OTP 21.1
- 1.7.4 OTP 21.2
- 1.8.1 OTP 21.1

My workaround is to ignore `:saxy` as a dependency in the dialyzer config. I'm using [dialyxir](https://github.com/jeremyjh/dialyxir) to run the dialyzer.

**Comment by qcam** [comment]
@the-mikedavis Thanks for reporting. Anyway I'm not experienced with dialyzer so not sure what I can do to help in this case (please let me know if you have any idea). Maybe you could try reporting to dialyzer repository?

**Comment by the-mikedavis** [comment]
Hi @qcam. Sorry to hit you with this weird issue and give you no leads (especially if you haven't used the dialyzer before).

The more I investigate, the more I imagine that this is an upstream Erlang issue related to [the bug report](https://bugs.erlang.org/browse/ERL-680).

I isolated `lib/saxy/parser/prolog.ex` into its own erlang module and I can pass the dialyzer (quickly) by commenting the equivalents of these four function clauses:

- [`prolog_pi_name/7` on line 351](https://github.com/qcam/saxy/blob/master/lib/saxy/parser/prolog.ex#L351)
- [`prolog_pi_name/7` on line 356](https://github.com/qcam/saxy/blob/master/lib/saxy/parser/prolog.ex#L356)
- [`dtd_pi_name/6` on line 493](https://github.com/qcam/saxy/blob/master/lib/saxy/parser/prolog.ex#L493)
- [`dtd_pi_name/6` on line 499](https://github.com/qcam/saxy/blob/master/lib/saxy/parser/prolog.ex#L498)

And these four are the only functions in the module to use the `is_name_char` guard. Oddly enough, functions with the `is_name_start_char` guard, which has many `orelse` and `andalso` clauses, are fine.

I think that to solve this in the short run, we might need to tweak the `is_name_char` guard because it has some awkward associativity rules. I think in the long run, the compiler should transform these cases a little differently from how it's described in [the patch here](https://github.com/erlang/otp/pull/1944/files).

I'll add more comments as I find things, submit a PR if I can get something working, and submit a bug report to Erlang if I can get something broken.

**Comment by qcam** [comment]
Thanks for the help with investigation. ❤️

> And these four are the only functions in the module to use the `is_name_char` guard. Oddly enough, functions with the `is_name_start_char` guard, which has many `orelse` and `andalso` clauses, are fine.

Actually `is_name_char` has a call to `is_name_start_char` (https://github.com/qcam/saxy/blob/master/lib/saxy/guards.ex#L38). Would that be the issue (that dialyzer does not work well when combining `or` with a macro)?

**Comment by the-mikedavis** [comment]
I can get the dialyzer to finish analyzing the whole project in about 6 minutes with these guards:

```elixir
  defguard is_ascii(charcode) when charcode <= 0x7F

  defguard is_whitespace(charcode) when charcode in [0xA, 0x9, 0xD, 0x20]

  defguard is_name_start_char(charcode)
  when charcode == ?: or
  # (charcode >= ?A and charcode <= ?Z) or
  # charcode == ?_ or
  # (charcode >= ?a and charcode <= ?z) or
  # (charcode >= 0xC0 and charcode <= 0xD6) or
  (charcode >= 0xD8 and charcode <= 0xF6) or
  (charcode >= 0xF8 and charcode <= 0x2FF)  or
  (charcode >= 0x370 and charcode <= 0x37D) or
  (charcode >= 0x37F and charcode <= 0x1FFF) or
  (charcode >= 0x200C and charcode <= 0x200D) or
  (charcode >= 0x2070 and charcode <= 0x218F)  or
  (charcode >= 0x2C00 and charcode <= 0x2FEF) or
  (charcode >= 0x3001 and charcode <= 0xD7FF) or
  (charcode >= 0xF900 and charcode <= 0xFDCF) or
  (charcode >= 0xFDF0 and charcode <= 0xFFFD) or
  (charcode >= 0x10000 and charcode <= 0xEFFFF)

  defguard is_name_char(charcode)
  when true
```

If I comment out another case of the `is_name_start_char`, it only takes 2 minutes. If I comment out 2 more cases of `is_name_start_char`, it only takes 40 seconds. 3 clauses and it takes 20 seconds.

So I think that the dialyzer can't analyze these efficiently. (It looks a little like exponential growth on the order of `orelse` clauses.)

If this was very high priority for you, I would recommend turning these guards into functions and using case statements. It's hard to say if it would be more efficient in the runtime or not. I'll try it out and see what happens.

Since the dialyzer isn't critical for this project and there's a workaround of ignoring the dependency, I think just submitting a bug report to Erlang will be suitable. I'll link to the bug report here once I've submitted it.

**Comment by the-mikedavis** [comment]
Ah so the untouched expansion hangs:

```erlang
prolog_pi_name(<<_charcode@1/utf8, _rest@1/bitstring>>,
               _more@1, _original@1, _pos@1, _state@1, _prolog@1,
               _len@1)
  when (((_charcode@1 >= 48 andalso _charcode@1 =< 57
          orelse
          _charcode@1 =:= 46 orelse
          _charcode@1 =:= 183 orelse _charcode@1 =:= 45)
         orelse
           ((((((((((((((_charcode@1 == 58 orelse
                         _charcode@1 >= 65 andalso _charcode@1 =< 90)
                        orelse _charcode@1 == 95)
                       orelse _charcode@1 >= 97 andalso _charcode@1 =< 122)
                      orelse _charcode@1 >= 192 andalso _charcode@1 =< 214)
                     orelse _charcode@1 >= 216 andalso _charcode@1 =< 246)
                    orelse _charcode@1 >= 248 andalso _charcode@1 =< 767)
                   orelse _charcode@1 >= 880 andalso _charcode@1 =< 893)
                  orelse _charcode@1 >= 895 andalso _charcode@1 =< 8191)
                 orelse _charcode@1 >= 8204 andalso _charcode@1 =< 8205)
                orelse _charcode@1 >= 8304 andalso _charcode@1 =< 8591)
               orelse
               _charcode@1 >= 11264 andalso _charcode@1 =< 12271)
              orelse
              _charcode@1 >= 12289 andalso _charcode@1 =< 55295)
             orelse
             _charcode@1 >= 63744 andalso _charcode@1 =< 64975)
            orelse
            _charcode@1 >= 65008 andalso _charcode@1 =< 65533)
         orelse
         _charcode@1 >= 65536 andalso _charcode@1 =< 983039)
        orelse _charcode@1 >= 768 andalso _charcode@1 =< 879)
       orelse
       _charcode@1 >= 8255 andalso _charcode@1 =< 8256 ->
    prolog_pi_name(_rest@1, _more@1, _original@1, _pos@1,
                   _state@1, _prolog@1,
                   _len@1 +
                       'Elixir.Saxy.Parser.Utils':compute_char_len(_charcode@1)).
```

But if we right-associate the `orelse` clauses, the computation is instant:

```erlang
prolog_pi_name(<<_charcode@1/utf8, _rest@1/bitstring>>,
         _more@1, _original@1, _pos@1, _state@1, _prolog@1,
         _len@1)
  when (_charcode@1 >= 48 andalso _charcode@1 =< 57)
       orelse (_charcode@1 =:= 46)
       orelse (_charcode@1 =:= 183)
       orelse (_charcode@1 =:= 45)
       orelse (_charcode@1 == 58)
       orelse (_charcode@1 >= 65 andalso _charcode@1 =< 90)
       orelse (_charcode@1 == 95)
       orelse (_charcode@1 >= 97 andalso _charcode@1 =< 122)
       orelse (_charcode@1 >= 192 andalso _charcode@1 =< 214)
       orelse (_charcode@1 >= 216 andalso _charcode@1 =< 246)
       orelse (_charcode@1 >= 248 andalso _charcode@1 =< 767)
       orelse (_charcode@1 >= 880 andalso _charcode@1 =< 893)
       orelse (_charcode@1 >= 895 andalso _charcode@1 =< 8191)
       orelse (_charcode@1 >= 8204 andalso _charcode@1 =< 8205)
       orelse (_charcode@1 >= 8304 andalso _charcode@1 =< 8591)
       orelse (_charcode@1 >= 11264 andalso _charcode@1 =< 12271)
       orelse (_charcode@1 >= 12289 andalso _charcode@1 =< 55295)
       orelse (_charcode@1 >= 63744 andalso _charcode@1 =< 64975)
       orelse (_charcode@1 >= 65008 andalso _charcode@1 =< 65533)
       orelse (_charcode@1 >= 65536 andalso _charcode@1 =< 983039)
       orelse (_charcode@1 >= 768 andalso _charcode@1 =< 879)
       orelse (_charcode@1 >= 8255 andalso _charcode@1 =< 8256) ->
    prolog_pi_name(_rest@1, _more@1, _original@1, _pos@1,
       _state@1, _prolog@1,
       _len@1 +
         'Elixir.Saxy.Parser.Utils':compute_char_len(_charcode@1)).
```

And [this PR](https://github.com/erlang/otp/pull/1944/files) is supposed to do right-association in the compiler. So it looks like this is really just a core Erlang problem!

**Comment by qcam** [comment]
@the-mikedavis 👍Thanks for the great investigation 💛.

I guess I'll close the issue then since there's no action to take on the library side.

**Comment by the-mikedavis** [comment]
https://github.com/erlang/otp/pull/2141 seems to work. Building OTP and elixir from source off of that gives me `7.5 sec` on the dialyzer 🎉 

I'll update when it gets merged into a release

**Comment by the-mikedavis** [comment]
PR merged and issue closed down! Should be in Erlang 21.3 but no release date is set yet.

**Comment by qcam** [comment]
@the-mikedavis 🎉awesome! Thank you for all the helps!
